### PR TITLE
Initial rows for CourseType, CourseRunType, Track, and Mode

### PR DESCRIPTION
**Pull request by Dillon-Dumesnil**
DISCO-1382

### CODE REVIEW

File: course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    #     help_text=_('Completion can grant credit toward an organization’s degree.'),
+    # )
+    # certificate_type = models.CharField(
+    #     max_length=64, choices=CertificateType, blank=True,
+    #     help_text=_('Certificate type granted if this mode is eligible for a certificate, or blank if not.'),
+    # )
+    # payee = models.CharField(
+    #     max_length=64, choices=PayeeType, default=PayeeType.Platform,
+    #     help_text=_('Who gets paid for the course? Platform is the site owner, Organization is the school.'),
+    # )
```
Comment (Dillon-Dumesnil) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: I'll remove all the comments after review. Just put them in there for now so it's easier to see what all the possible fields are for these models. The fields are below the actual model variable declaration.

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -259,7 +259,7 @@ class Mode(TimeStampedModel):
         help_text=_('Certificate type granted if this mode is eligible for a certificate, or blank if not.'),
     )
     payee = models.CharField(
-        max_length=64, choices=PayeeType, default=PayeeType.Platform,
+        max_length=64, choices=PayeeType, default=PayeeType.Platform, blank=True,
```
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py]: It's possible it might be nice to add an `@property` like `def is_paid():` or something to check a blank value here (as we have for certificate type). Not a blocker for this PR, just something to think about as we use these.

File: course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        'name': 'Verified',
+        'is_id_verified': True,
+        'certificate_type': CertificateType.Verified,
+    }),
+)
+# The slugs are identical across Mode and SeatType
+TRACK_MODE_AND_SEAT_SLUGS = ('audit', 'credit', 'professional', 'verified',)
+COURSE_RUN_TYPE_NAMES_AND_TRACK_SLUGS = (
+    ('Audit Only', ['audit']),
+    ('Credit', ['audit', 'credit', 'verified']),
```
Comment (jmyatt) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: Julia confirmed that currently, all runs for Credit courses have all 3 of these track types

File: course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+MODE_SLUGS_AND_DEFAULTS = (
+    ('audit', {
+        'name': 'Audit',
+        'payee': '',
+    }),
+    ('credit', {
+        'name': 'Credit',
+        'is_id_verified': True,
+        'is_credit_eligible': 1,
+        'certificate_type': CertificateType.Credit,
```
Comment (mikix) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: Ideally we don't have to hard-import CetificateType. So maybe hardcode the string here, or move these definitions into the method where you can do like `apps.get_model`

File: course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+from django.db import migrations, models
+
+from course_discovery.apps.course_metadata.choices import CertificateType, PayeeType
+
+SEAT_TYPES = ('Audit', 'Credit', 'Professional', 'Verified',)
+# Contains the slug (unique identifier) and any fields that are not the default values
+MODE_SLUGS_AND_DEFAULTS = (
+    ('audit', {
+        'name': 'Audit',
+        'payee': '',
```
Comment (mikix) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: Is it weird that payee doesn't default to this?

File: course_discovery/apps/course_metadata/models.py (hunk 2 above)
Comment (Dillon-Dumesnil) [reply course_discovery/apps/course_metadata/models.py]: Yeah, I feel like we should maybe wait and see if that is needed before adding it. I'm not sure of the use case yet.

File: course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        mode = Mode.objects.get(slug=slug)
+        seat_type = SeatType.objects.get(slug=slug)
+        Track.objects.update_or_create(mode=mode, seat_type=seat_type)
+    for name, track_slugs in COURSE_RUN_TYPE_NAMES_AND_TRACK_SLUGS:
+        tracks = []
+        for slug in track_slugs:
+            mode = Mode.objects.get(slug=slug)
+            seat_type = SeatType.objects.get(slug=slug)
+            tracks.append(Track.objects.get(mode=mode, seat_type=seat_type))
+        run_type, __ = CourseRunType.objects.update_or_create(name=name)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: Why a dunder? One should be enough?

File: course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        run_type.tracks.set(tracks)
+    for course_name, seat_type_slugs, run_type_names in COURSE_TYPE_NAMES_AND_SEAT_TYPE_SLUGS_AND_RUN_TYPE_NAMES:
+        entitlement_types = []
+        for slug in seat_type_slugs:
+            entitlement_types.append(SeatType.objects.get(slug=slug))
+        run_types = []
+        for run_name in run_type_names:
+            run_types.append(CourseRunType.objects.get(name=run_name))
```
Comment (mikix) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: Maybe:
```
entitlement_types = {SeatType.objects.get(slug=slug) for slug in seat_type_slugs}
run_types = {CourseRunType.objects.get(name=run_name) for run_name in run_names}
```

Comment (mikix) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: When grabbing CourseRunType by name, maybe do a `filter().first()`, since technically there could be multiple if someone made one after migration that made the model, but before this one...

I guess similar problem right below on the CourseType update_or_create... I dunno what to do about it.  I guess you could keep a live mapping of created-UUIDs to the names assigned. But meh.

Maybe just leave it all as-is.

**Review by mikix — APPROVED** [review]
Nice.

File: course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py [hunk 8]
```diff
@@ -23,16 +23,14 @@ def setUp(self):
...
-                                                          entitlement_types=[self.verified_seat_type],
-                                                          course_run_types=[self.audit_run_type, self.va_run_type])
+        self.audit_mode = Mode.objects.get(slug=Seat.AUDIT)
+        self.verified_mode = Mode.objects.get(slug=Seat.VERIFIED)
+        self.audit_track = Track.objects.get(seat_type=self.audit_seat_type, mode=self.audit_mode)
+        self.verified_track = Track.objects.get(seat_type=self.verified_seat_type, mode=self.verified_mode)
+        self.audit_run_type = CourseRunType.objects.get(name='Audit Only')
+        self.va_run_type = CourseRunType.objects.get(name='Verified and Audit')
+        self.va_course_type = CourseType.objects.get(name='Verified and Audit')
+        self.audit_course_type = CourseType.objects.get(name='Audit Only')
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py]: This kinda thing is annoying...  We should maybe add a slug field rather than/additionally to the model? Or make `name` unique, but that feels wrong to me.

File: course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py [hunk 9]
```diff
@@ -202,6 +200,9 @@ def test_matches_earliest_course_type(self):
         self.run_command()
         self.assertEqual(self.course.type, self.va_course_type)
 
+        # This messes up this test due to Verified and Audit being a subset of Credit
+        # and Credit being created prior to 'Second'
+        CourseType.objects.get(name='Credit').delete()
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py]: I would move this to the top of the test function, so it's true for all the mini-tests I do.

File: course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py [hunk 10]
```diff
@@ -223,39 +224,32 @@ def test_one_mismatched_run_ruins_whole_thing(self):
     def test_by_org(self):
         self.run_command(orgs=[self.org2])
         self.assertIsNone(self.course.type)
-        self.assertEqual(self.course2.type, self.va_course_type)
+        self.assertEqual(self.course2.type, self.audit_course_type)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py:227]: Wait, why these changes? Course has runs with verified seats, yah?

File: course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py [hunk 11]
```diff
@@ -223,39 +224,32 @@ def test_one_mismatched_run_ruins_whole_thing(self):
...
         self.assertEqual(draft_course.entitlements.first().mode.slug, Seat.AUDIT)  # made a draft audit entitlement
         self.assertIsNone(self.course2.entitlements.first())  # but official is still empty
 
-        # Will fail without an audit-only CourseType
-        self.run_command(courses=[self.course2], fails=draft_course)  # self.course2 passes with V&A type
-
-        # Now make the missing audit-only CourseType and try again
-        audit_course_type = factories.CourseTypeFactory(name='Audit Only', entitlement_types=[],
-                                                        course_run_types=[self.audit_run_type])
-
```
Comment (mikix) [inline course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py]: I would prefer deleting the audit course type first before this stuff (and credit type?), so we can still test that we fail without audit-only.

File: course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py (hunk 10 above)
Comment (Dillon-Dumesnil) [reply course_discovery/apps/course_metadata/management/commands/tests/test_backpopulate_course_type.py:227]: Nah, only audit

**Review by mikix — APPROVED** [review]
Nice! Slugs do look nice.

File: course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+COURSE_RUN_TYPE_NAMES_SLUGS_AND_TRACK_SLUGS = (
+    ('Audit Only', 'audit', ['audit']),
+    ('Professional Only', 'professional', ['professional']),
+    ('Verified and Audit', 'verified-and-audit', ['audit', 'verified']),
+    ('Credit', 'credit', ['audit', 'credit', 'verified']),
+)
+COURSE_TYPE_NAMES_SLUGS_SEAT_TYPE_SLUGS_AND_RUN_TYPE_NAMES = (
+    ('Audit Only', 'audit', ['audit'], ['Audit Only']),
+    ('Professional Only', 'professional', ['professional'], ['Professional Only']),
+    ('Verified and Audit', 'verified-and-audit', ['verified'], ['Audit Only', 'Verified and Audit']),
```
Comment (mikix) [inline course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: Could go `verified` instead? Like, `credit` doesn't include other subtypes too. But super nit.

Comment (Dillon-Dumesnil) [reply course_discovery/apps/course_metadata/migrations/0204_create_initial_coursetypes.py]: I wanted to keep this that way in case we introduce a verified only version one day. But don't disagree that including the subtypes could be problematic as we continue to increase.

`masters-verified-audit` here we come
